tienPh commit: contract Marketplace.sol code auction

Niet-Pham opened the pull request:
-create: Marketplace.sol. code functions auction

Niet-Pham committed:
tienPh commit: contract Marketplace.sol code auction

Niet-Pham committed:
thêm thuộc tính view cho các function private và xóa "pragma experimental ABIEncoderV2;"

Niet-Pham committed:
contract NginNFT:
-thêm mapping(tokenId=>tokenHash) tokenHashById
-function mint thêm parameter string tokenHash và khi mint thì cho vào mapping tokenHashById
- overrride function tokenURI(tokenId) để trả về baseURI/tokenHash

xieuxaythit commented on contracts/NginNFT.sol:
```diff
@@ -16,13 +17,16 @@ contract NginNFT is ERC721, ERC721Enumerable, ERC721Burnable, Ownable {
 
     constructor(string memory _uri) ERC721("NginNFT", "NGNFT") {
-        _setBaseURI(_uri);
+        _tokenBaseUri = _uri;
```
@Niet-Pham chỗ này ko cần hàm _setBaseURI()?

Niet-Pham replied:
hình như chỗ này trong code của em nó cũng k để setBaseURi a ạ.

Niet-Pham committed:
Marketplace.sol:
-Ghép code đấu giá với hợp đồng mua bán của namph
-Fix code hợp đồng mua bán của namph

xieuxaythit commented on contracts/NginNFT.sol:
```diff
@@ -16,13 +17,16 @@ contract NginNFT is ERC721, ERC721Enumerable, ERC721Burnable, Ownable {
 
     string private _tokenBaseUri;
+    
+    mapping(uint => string) tokenHashById;
```
``` Suggestion
    mapping(uint => bytes32) tokenHashById;
```
Chuyển về bytes32 sẽ tiết kiệm bộ nhớ hơn

Niet-Pham committed:
-contract NFT: sửa khi deploy gọi hàm setBaseUri như theo comment của a Hoàng a
-contract Marketplace:
  +Import và implêmnt ERC721Holder
  +sửa khi tạo phiên đấu giá/mua bán thì chuyển token vào contract
  + thêm chức năng tại hàm kết thúc phiên đấu giá: kiểm tra xem có ai đấu giá k. nếu k thì chuyển token trả về người bán
  + Thêm enum enum enTrangThaiHopDong{ DangBan, DaBan, HuyBan}
  + thay đổi struc HopDongMuaBan. thay (bool HoatDong) thành (enTrangThaiHopDong TrangThaiHopDong)
  +thêm hàm hủy bán hàng cho luồng mua bán thông thường: nếu hàng vẫn đang có trạng thái là đang bán. Người bán có thể hủy để lấy lại token từ contract và hợp đồng mua bán này chuyển sang trang thái hủy bán
